### PR TITLE
MSI connectivity feature store

### PR DESCRIPTION
**Pull request by jedheaj314**
Following tasks have been completed under this PR

-  Create a new AD group for ${var.suffix} flowehr-apps 
-  Create a new AD group for ${var.suffix} flowehr-developers
-  Assign both groups read permissions on the database using the create_sql_user.py script
-  Output the AD group principal ids from the module within apps/app/platform.tf
-  Add the web app MSI (azurerm_linux_web_app.app.identity[0].principal_id) to the flowehr-apps AD group
-  Change the apps.yaml config so owners and contributors is a map of gh_username:email
-  Add the emails from contributors to the flower-developers group

### CODE REVIEW

File: infrastructure/transform/feature-data-store.tf [hunk 1]
```diff
@@ -216,3 +216,41 @@ resource "azurerm_private_endpoint" "sql_server_features_pe" {
...
+  security_enabled = true
+}
+
+resource "azuread_group" "ad_group_data_scientists" {
+  display_name     = "${var.naming_suffix} flowehr-data-scientists"
+  owners           = [data.azurerm_client_config.current.object_id]
+  security_enabled = true
+}
+
+output "apps_ad_group_display_name" {
```
Comment (jjgriff93) [inline infrastructure/transform/feature-data-store.tf:238]: Suggest moving these to `outputs.tf`

File: infrastructure/transform/feature-data-store.tf [hunk 2]
```diff
@@ -216,3 +216,41 @@ resource "azurerm_private_endpoint" "sql_server_features_pe" {
...
+  description = "Apps AD group display name"
+  value       = azuread_group.ad_group_apps.display_name
+}
+
+output "developers_ad_group_display_name" {
+  description = "Apps AD group display name"
```
Comment (jjgriff93) [inline infrastructure/transform/feature-data-store.tf:244]:
```suggestion
  description = "Developers AD group display name"
```

File: infrastructure/transform/feature-data-store.tf [hunk 3]
```diff
@@ -216,3 +216,41 @@ resource "azurerm_private_endpoint" "sql_server_features_pe" {
...
+  value       = azuread_group.ad_group_developers.display_name
+}
+
+output "apps_ad_group_principal_id" {
+  description = "Apps AD group principal id "
```
Comment (jjgriff93) [inline infrastructure/transform/feature-data-store.tf:249]:
```suggestion
  description = "Apps AD group principal id"
```

File: infrastructure/transform/feature-data-store.tf [hunk 4]
```diff
@@ -216,3 +216,41 @@ resource "azurerm_private_endpoint" "sql_server_features_pe" {
...
+  value       = azuread_group.ad_group_apps.object_id
+}
+
+output "developers_ad_group_principal_id" {
+  description = "Developers AD group principal id "
```
Comment (jjgriff93) [inline infrastructure/transform/feature-data-store.tf:254]:
```suggestion
  description = "Developers AD group principal id"
```

File: apps/terragrunt.hcl [hunk 5]
```diff
@@ -82,13 +86,18 @@ inputs = {
...
   core_log_analytics_name = dependency.core.outputs.core_log_analytics_name
 
-  transform_feature_store_server_name = dependency.transform.outputs.feature_store_server_name
-  transform_feature_store_db_name     = dependency.transform.outputs.feature_store_db_name
+  transform_feature_store_server_name  = dependency.transform.outputs.feature_store_server_name
+  transform_feature_store_db_name      = dependency.transform.outputs.feature_store_db_name
+  transform_apps_ad_group_display_name = dependency.transform.outputs.apps_ad_group_display_name
+  transform_developers_ad_group_display_name = dependency.transform.outputs.developers_ad_group_display_name
+  transform_apps_ad_group_principal_id = dependency.transform.outputs.apps_ad_group_principal_id
+  transform_developers_ad_group_principal_id = dependency.transform.outputs.developers_ad_group_principal_id
```
Comment (jjgriff93) [inline apps/terragrunt.hcl:94]:
```suggestion
  transform_feature_store_server_name                = dependency.transform.outputs.feature_store_server_name
  transform_feature_store_db_name                      = dependency.transform.outputs.feature_store_db_name
  transform_apps_ad_group_display_name           = dependency.transform.outputs.apps_ad_group_display_name
  transform_developers_ad_group_display_name = dependency.transform.outputs.developers_ad_group_display_name
  transform_apps_ad_group_principal_id               = dependency.transform.outputs.apps_ad_group_principal_id
  transform_developers_ad_group_principal_id     = dependency.transform.outputs.developers_ad_group_principal_id
```

Comment (jjgriff93) [reply apps/terragrunt.hcl:94]: Check this alignment in VS Code as the GitHub editor is not ideal for this! Basically all the = signs should align in all HCL files not just TF ones

File: apps/terragrunt.hcl [hunk 6]
```diff
@@ -82,13 +86,18 @@ inputs = {
...
 
   serve_app_service_plan_name = dependency.serve.outputs.app_service_plan_name
   serve_acr_name              = dependency.serve.outputs.acr_name
   serve_cosmos_account_name   = dependency.serve.outputs.cosmos_account_name
   serve_webapps_subnet_id     = dependency.serve.outputs.webapps_subnet_id
 
+
```
Comment (jjgriff93) [inline apps/terragrunt.hcl:101]:
```suggestion
```

File: apps/terragrunt.hcl [hunk 7]
```diff
@@ -60,6 +60,10 @@ dependency "transform" {
   mock_outputs = {
     feature_store_server_name = "transform_feature_store_server_name"
     feature_store_db_name     = "transform_feature_store_db_name"
+    apps_ad_group_display_name = "transform_apps_ad_group_display_name"
```
Comment (jjgriff93) [inline apps/terragrunt.hcl:63]: Align this block's equals signs

File: apps/app/variables.tf [hunk 8]
```diff
@@ -78,12 +78,28 @@ variable "github_owner" {
...
+
+variable "developers_ad_group_principal_id" {
+  type = string
+}
+
 variable "app_config" {
   type = object({
     name         = string
     description  = string
     owners       = set(string)
```
Comment (jjgriff93) [inline apps/app/variables.tf:101]: `map` here as well

File: apps/app/platform.tf [hunk 9]
```diff
@@ -158,4 +158,15 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
   role_definition_name = azurerm_role_definition.slot_swap[0].name
 }
 
+resource "azuread_group_member" "web_app_in_app_ad_group" {
```
Comment (jjgriff93) [inline apps/app/platform.tf:161]: Keep it simple :) 
```suggestion
resource "azuread_group_member" "webapp" {
```

File: apps/app/platform.tf [hunk 10]
```diff
@@ -158,4 +158,15 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+  group_object_id  = var.apps_ad_group_principal_id
+  member_object_id = azurerm_linux_web_app.app.identity[0].principal_id
+}
+
+resource "azuread_group_member" "contributors_in_dev_ad_group" {
```
Comment (jjgriff93) [inline apps/app/platform.tf:166]:
```suggestion
resource "azuread_group_member" "developers" {
```

**Review by jjgriff93 — COMMENTED** [review]
Good work, some comments and suggestions - you'll also want to merge in the latest changes as there's some conflicts so give me a shout if you need help with this!

File: infrastructure/transform/feature-data-store.tf (hunk 1 above)
Comment (jedheaj314) [reply infrastructure/transform/feature-data-store.tf]: Done 8cbba0c21ea5058129e27fc2f66482b6901cafce

File: apps/app/variables.tf (hunk 8 above)
Comment (jedheaj314) [reply apps/app/variables.tf]: Done 7d9c95487fc4150de9d1797cd6d29c9d6fd99456

File: apps/terragrunt.hcl (hunk 5 above)
Comment (jedheaj314) [reply apps/terragrunt.hcl]: Done 2d55b388b080110b76af6be5cbc5da3fda0343b8

File: infrastructure/transform/feature-data-store.tf (hunk 3 above)
Comment (jedheaj314) [reply infrastructure/transform/feature-data-store.tf]: Done 21b63ce4a07960af4d66c28006b6d886afd88874

File: infrastructure/transform/feature-data-store.tf (hunk 2 above)
Comment (jedheaj314) [reply infrastructure/transform/feature-data-store.tf]: Done 2cc7a8fd837987a27426e274d67d3817e44657af

File: apps/app/platform.tf [hunk 11]
```diff
@@ -158,4 +158,15 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+  member_object_id = azurerm_linux_web_app.app.identity[0].principal_id
+}
+
+resource "azuread_group_member" "developers" {
+  for_each         = data.azuread_user.contributors_ids
+  group_object_id  = var.developers_ad_group_principal_id
+  member_object_id = each.value.object_id
+}
+
 // TODO: once Feature Store SQL SPN stuff is in, add connection from App Service here
```
Comment (jjgriff93) [inline apps/app/platform.tf:172]:
```suggestion
```

**Review by damoodamoo — CHANGES_REQUESTED** [review]
1 spanner-throwing thought - I think we might cause ourselves a problem by directly assigning the contributors to the App Developers AAD group. 

If I am a contributor to 2x apps, and one gets removed, I'm pretty sure terraform will remove the role assignment, which would remove me from the group - even though i'm a contributor to another app. 

I would suggest we create a new AAD group per app, populate that with all the developers, and then add that new AAD group into the App Developers group - so removing an app would remove the AAD group specific for that app, and nothing overlaps.

**Comment by jedheaj314** [comment]
> 1 spanner-throwing thought - I think we might cause ourselves a problem by directly assigning the contributors to the App Developers AAD group.
> 
> If I am a contributor to 2x apps, and one gets removed, I'm pretty sure terraform will remove the role assignment, which would remove me from the group - even though i'm a contributor to another app.
> 
> I would suggest we create a new AAD group per app, populate that with all the developers, and then add that new AAD group into the App Developers group - so removing an app would remove the AAD group specific for that app, and nothing overlaps.

Done 55b131849a7d5a5e28e15408d73acddde3278ec7

File: apps/app/platform.tf [hunk 12]
```diff
@@ -224,4 +224,23 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
 }
 
-// TODO: once Feature Store SQL SPN stuff is in, add connection from App Service here
+resource "azuread_group_member" "webapp" {
+  group_object_id  = var.apps_ad_group_principal_id
+  member_object_id = azurerm_linux_web_app.app.identity[0].principal_id
+}
+
+resource "azuread_group" "app" {
+  display_name     = "${var.naming_suffix} app"
```
Comment (t-young31) [inline apps/app/platform.tf:233]: should be `app_id` I think?

File: apps/app/platform.tf [hunk 13]
```diff
@@ -224,4 +224,23 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+}
+
+resource "azuread_group_member" "app_developers" {
+  for_each         = data.azuread_user.contributors_ids
+  group_object_id  = azuread_group.app.id
+  member_object_id = each.value.object_id
+}
+
+resource "azuread_group_member" "developers" {
+  group_object_id  = var.developers_ad_group_principal_id
```
Comment (t-young31) [inline apps/app/platform.tf:244]: seems odd to me that an app goes in a developers group(?)

File: infrastructure/transform/feature-data-store.tf [hunk 14]
```diff
@@ -219,3 +219,21 @@ resource "azurerm_private_endpoint" "sql_server_features_pe" {
...
+  display_name     = "${var.naming_suffix} flowehr-developers"
+  owners           = [data.azurerm_client_config.current.object_id]
+  security_enabled = true
+}
+
+resource "azuread_group" "ad_group_data_scientists" {
+  display_name     = "${var.naming_suffix} flowehr-data-scientists"
+  owners           = [data.azurerm_client_config.current.object_id]
+  security_enabled = true
+}
```
Comment (t-young31) [inline infrastructure/transform/feature-data-store.tf:239]: if these groups get used in transform+apps+tre then I would make them in core?

File: scripts/sql/create_sql_user.py [hunk 15]
```diff
@@ -14,42 +14,51 @@
...
+users_to_create = os.environ.get("USERS_TO_CREATE")
 
 
 def create_con_str(db: str) -> str:
     driver = "{ODBC Driver 18 for SQL Server}"
     return f"DRIVER={driver};SERVER={server};DATABASE={db};ENCRYPT=yes;Authentication=ActiveDirectoryServicePrincipal;UID={client_id};PWD={client_secret}"  # noqa: E501
 
 
+# Map users
+users_data = json.loads(users_to_create)
```
Comment (t-young31) [inline scripts/sql/create_sql_user.py:32]: nit: I would have the json as a list then you can just

```python
users = json.loads(users_to_create)
```
which is cleaner, imo anyway

**Review by t-young31 — COMMENTED** [review]
NIce! have added some comments

File: apps/app/platform.tf (hunk 12 above)
Comment (jjgriff93) [inline apps/app/platform.tf]:
```suggestion
resource "azuread_group" "developers_of_this_app" {
  display_name     = "${var.app_id} app developers"
```

File: apps/app/platform.tf [hunk 16]
```diff
@@ -224,4 +224,23 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+  group_object_id  = var.apps_ad_group_principal_id
+  member_object_id = azurerm_linux_web_app.app.identity[0].principal_id
+}
+
+resource "azuread_group" "app" {
+  display_name     = "${var.naming_suffix} app"
+  security_enabled = true
+}
+
+resource "azuread_group_member" "app_developers" {
```
Comment (jjgriff93) [inline apps/app/platform.tf:237]:
```suggestion
resource "azuread_group_member" "add_each_developer_to_group" {
```

File: apps/app/platform.tf [hunk 17]
```diff
@@ -224,4 +224,23 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+  for_each         = data.azuread_user.contributors_ids
+  group_object_id  = azuread_group.app.id
```
Comment (jjgriff93) [inline apps/app/platform.tf:239]:
```suggestion
  group_object_id  = azuread_group.developers_of_this_app.id
```

File: apps/app/platform.tf [hunk 18]
```diff
@@ -224,4 +224,23 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+  member_object_id = each.value.object_id
+}
+
+resource "azuread_group_member" "developers" {
```
Comment (jjgriff93) [inline apps/app/platform.tf:243]:
```suggestion
resource "azuread_group_member" "add_developers_of_app_group_to_parent" {
```

File: apps/app/platform.tf [hunk 19]
```diff
@@ -224,4 +224,23 @@ resource "azurerm_role_assignment" "webapp_sp_slot_swap" {
...
+  group_object_id  = var.developers_ad_group_principal_id
+  member_object_id = azuread_group.app.id
```
Comment (jjgriff93) [inline apps/app/platform.tf:245]:
```suggestion
  member_object_id = azuread_group.developers_of_this_app.id
```

File: infrastructure/transform/feature-data-store.tf (hunk 14 above)
Comment (jjgriff93) [reply infrastructure/transform/feature-data-store.tf:239]: Not sure I agree, think it should happen just after database creation so seems like the right place

Comment (t-young31) [reply infrastructure/transform/feature-data-store.tf:239]: is their only purpose going to be for feature store access? a group full of app devs would be useful for other things(?)
(feel free to ignore - don’t feel strongly)

Comment (jjgriff93) [reply infrastructure/transform/feature-data-store.tf:239]: Potentially - I reckon we move when that becomes the case, right now there aren't any other use cases I'm aware of, but fair point!

File: scripts/sql/create_sql_user.py [hunk 20]
```diff
@@ -14,42 +14,51 @@
...
 cnxn = pyodbc.connect(create_con_str(database))
 cursor = cnxn.cursor()
 
-query = f"""
-IF NOT EXISTS(SELECT principal_id FROM sys.database_principals WHERE name = '{login_to_create}') BEGIN
-    CREATE USER [{login_to_create}] FROM EXTERNAL PROVIDER
-    EXEC sp_addrolemember 'db_owner', [{login_to_create}]
-END
-"""  # noqa: E501
+for user in users_to_create:
```
Comment (t-young31) [inline scripts/sql/create_sql_user.py:52]: think this should be 
```suggestion
for user in users:
```

**Comment by jedheaj314** [comment]
> Good work! Some naming suggestions to make things clearer, and I think Tom's point about the json list is spot on, would be cleaner

Done abaab4a5441e8ff1f76d329c97bb917449175f3b

File: scripts/sql/create_sql_user.py [hunk 21]
```diff
@@ -14,42 +14,51 @@
...
-# connect to target feature database to create user + assign as dbo
 cnxn = pyodbc.connect(create_con_str(database))
 cursor = cnxn.cursor()
 
-query = f"""
-IF NOT EXISTS(SELECT principal_id FROM sys.database_principals WHERE name = '{login_to_create}') BEGIN
-    CREATE USER [{login_to_create}] FROM EXTERNAL PROVIDER
-    EXEC sp_addrolemember 'db_owner', [{login_to_create}]
-END
-"""  # noqa: E501
```
Comment (jjgriff93) [inline scripts/sql/create_sql_user.py:52]: Looks like this is still incorrect as per Tom’s comment - this needs to be ‘users’ as well?

File: scripts/sql/create_sql_user.py (hunk 20 above)
Comment (jjgriff93) [reply scripts/sql/create_sql_user.py:52]: Looks like this still needs to be resolved @jedheaj314

**Review by t-young31 — APPROVED** [review]

**Pull request closed without merging**